plugin/src/main/java/com/denizenscript/denizen/utilities/Utilities.java

```diff
@@ -582,29 +582,45 @@ public static ElementTag enumLikeToLegacyElement(Object val) {
         throw new UnsupportedOperationException("Cannot get legacy name element, value isn't an enum: " + val);
     }
 
-    // TODO: need proper input backsupport, see https://discord.com/channels/315163488085475337/1011496047811506227/1301272242386370580
-    public static <T> T elementToEnumlike(ElementTag element, Class<T> type) {
-        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_21) && Keyed.class.isAssignableFrom(type)) {
-            return (T) Bukkit.getRegistry((Class<? extends Keyed>) type).get(parseNamespacedKey(element.asString()));
+    public static <T extends Keyed> T elementToEnumlike(ElementTag element, Class<T> type) {
+        return elementToEnumlike(element, type, true);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Keyed> T elementToEnumlike(ElementTag element, Class<T> type, boolean showWarning) {
+        if (NMSHandler.getVersion().isAtMost(NMSVersion.v1_20)) {
+            return (T) element.asEnum((Class<? extends Enum<?>>) type);
+        }
+        Registry<T> registry = Bukkit.getRegistry(type);
+        T value = registry.get(parseNamespacedKey(element.asString()));
+        if (!Settings.cache_legacySpigotNamesSupport || value != null) {
+            return value;
+        }
+        String updatedName = NMSHandler.instance.updateLegacyName(type, element.asString());
+        if (CoreUtilities.equalsIgnoreCase(element.asString(), updatedName)) {
+            return null;
+        }
+        if (showWarning) {
+            BukkitImplDeprecations.oldSpigotNames.warn();
         }
-        return (T) element.asEnum((Class<? extends Enum>) type);
+        return registry.get(parseNamespacedKey(updatedName));
     }
 
-    public static <T> T findBestEnumlike(Class<T> type, String... names) {
+    public static <T extends Keyed> T findBestEnumlike(Class<T> type, String... names) {
         for (String name : names) {
-            T val = elementToEnumlike(new ElementTag(name), type);
+            T val = elementToEnumlike(new ElementTag(name), type, false);
             if (val != null) {
                 return val;
             }
         }
         return null;
     }
 
-    public static boolean matchesEnumlike(ElementTag element, Class<?> type) {
-        return elementToEnumlike(element, type) != null;
+    public static boolean matchesEnumlike(ElementTag element, Class<? extends Keyed> type) {
+        return elementToEnumlike(element, type, false) != null;
     }
 
-    public static boolean requireEnumlike(Mechanism mechanism, Class<?> type) {
+    public static boolean requireEnumlike(Mechanism mechanism, Class<? extends Keyed> type) {

```

Comment:
the lack of `Keyed` is intentional, these methods are meant to support non-keyed enums too (eg legacy code that hasn't been converted to keyed registry stuff yet) (not that it matters much in practice since generic types in Java literally do nothing)

Reply:
Is that actually useful though? most relevant things were already `Keyed` even before getting properly converted (https://helpch.at/docs/1.17.1/org/bukkit/entity/Villager.Type.html for example), and even things that are still an enum (I think `PotionType` will need this for example? working on that PR separately though) are `Keyed` and have a registry.

Reply:
`Is that actually useful though?` asking backwards.

This PR is making a change to existing code. Is the change actually useful? Why is a new requirement for inputs being introduced? This is a util function that supported non-keyed enums before, and now it doesn't. Why? Why not just leave it with the support it had? If you remove support for things just because you don't see why it's there... you're gonna find out why it was there the hard way 3 months from now when somebody submits a bug report. (ie, Chesterton’s Fence)

Reply:
Well, the method is specifically `EnumLike`, a util for working with things that got updated and need compatibility across versions - if something is still fully on the legacy system and is an enum with no registry no keys and no renames, it shouldn't be using these methods as it doesn't need any form of special updating logic.
The method is also specifically built around that currently, I.e. if there's an enum that had renames but doesn't have a registry it can't use this method to support legacy names.

If you meant that regardless of the backsupport bit (as in, just using this method instead of `#asEnum` and still having the backsupport only work for registry stuff) then I guess that can be added, but I don't think there's much reason to have that extra logic?

Reply:
Yes, regardless of backsupport stuff. You're adding the backsupport stuff in this PR, but these methods predate this PR. These are existing general purpose utility functions.

`Enumlike` means "like an enum". Enums are like an enum. These functions are intended to be a universable callable that works with enums, registries, not-yet-converted enums that could be registries, and converted enums-to-registries, all equally well.
These should be safe to call on anything that is "like an enum" without having to think through exactly whether it's the most proper method or if there's technically a more direct call.

One thing I wanted to do but haven't bothered yet, is: use coreimpl calls to pass the core element enum handlers back to the impl to automatically handle all enumlikes with the core enum handles (ie enable `asEnum` and all to be universal automatically). Splitting the call to required keyed would make that harder to implement.

-----

But, again: unless you have a reason to require `Keyed`... why are you adding that requirement? If there was a good reason to add that requirement, we could go validate that nothing not Keyed ever goes in, and deal with the nuisance of the limitation in the future only once it gets in the way... *if* there was a good reason.

To break it down into logical principles of software design:
- "If it ain't broke, don't fix it" - why are you changing something? You need a reason, don't change for the sake of change.
- "Chesterton's Fence" - if you don't know why something is the way it is, don't mess with it until you do know.
- General code should always err on the side of overinclusion - unless there's a reason to restrict code usability, don't restrict it. Otherwise you're just being a nuisance for future developers who want to use it. You don't have to code in support for every possible idea somebody might have, but you shouldn't code in walls in the way of things.